### DISCUSSION TITLE
Window Refocus Refetching and Optimistic Updates

### DISCUSSION THREAD

**Discussion by jcmnunes** [TanStack/query#439] [General]
I think Window Refocus Refetching and Optimistic Updates are two great features of this library and I'd like to keep using both. However, I'm having a hard time trying to make them work together.

The scenario is this:

1. We have an optimistic mutation in the app
2. The user refocus the app and immediately triggers that optimistic update
3. Since we are refetching on window focus, our optimistic update gets reverted
4. Later, when the optimistic mutation resolves, we see another state change

![optimistic-update](https://user-images.githubusercontent.com/13286745/80958253-9cca4c80-8dfc-11ea-9ca0-df5542856275.gif)

I'm wondering, is there an easy way to improve this behavior?

**Comment by tannerlinsley** [comment]
I think this is something we should discuss deeper. The issue is definitely that a "refetch" starts happening before the optimistic update happens (or even right after it), and it brings in old data **after** the optimistic update. Maybe what we need to investigate is a way to "cancel" an outgoing query when an optimistic update is done. Would that potentially fix this issue?

**Reply by jcmnunes** [reply]
Hey @tannerlinsley , first of all, thank you for this great library. I've been using it lately and it has been a great experience so far 🙇 

Regarding this "issue", I agree with you. If somehow, when doing an optimistic update, we are able to cancel any outgoing query, that would be a viable fix.

The problem is that we might not want to cancel all outgoing queries, just the ones related to our mutation.

I would say that most of the time, the queries we might want to cancel are exactly the ones we refetch inside the mutation's `onSettled` handler.

**Comment by jcmnunes** [comment]
@tannerlinsley , I see you recently added a new method `queryCache.cancelQueries` that when used in Optimistic Updates, like it is described in the Readme [here](https://github.com/tannerlinsley/react-query#optimistic-updates-with-automatic-rollback-for-failed-mutations), can potentially fix this issue. What do you think?

On that note, is it production-ready already? I'm asking because there is no documentation for that method in the [API section](https://github.com/tannerlinsley/react-query#api). Also, there is no TypeScript support yet ([index.d.ts](https://github.com/tannerlinsley/react-query/blob/master/types/index.d.ts)).

I can open a PR if you want to fix both issues.

Thank you in advance 🙏

**Reply by tannerlinsley** [reply]
Yes it should be production ready. If you would like to add those docs/types, then I would gladly merge that PR :)

**Reply by jcmnunes** [reply]
Awesome, I will open the PR soon.
